### PR TITLE
Support unloadability in `Microsoft.CSharp`.

### LINKED ISSUE #71629: AssemblyLoadContext never get unloaded if there's a dynamic operation
### Description

There's an assembly with below code :
```c#
   dynamic eo = new ExpandoObject();
   eo.abc = 123; // this line will cause the issue
```
If I load the assembly in the `AssemblyLoadContext` with `isCollectible = true`, than this load context will never get unloaded via below methods:
```c#
loadContext.Unload();
GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();
```

I think there're some dynamic generated objects still reference the load context. But I can't find anyway to release them.

### Reproduction Steps

Just put the below code in a console app (.Net 5 or 6), and add `microsoft.codeanalysis.csharp.scripting` package will re-produce the issue.    
And you can find the output of the count always increase.

```c#
public class Program
{
    public static async Task Main(string[] args)
    {
       // ==== prepare the assembly ====
        string code = @"dynamic eo = new ExpandoObject(); 
                        eo.abc = 123; // comment out this line will fix the issue
                        Console.WriteLine(System.AppDomain.CurrentDomain.GetAssemblies().Count());";
        ScriptOptions so = ScriptOptions.Default
            .AddImports("System", "System.Linq", "System.Dynamic")
            .AddReferences("System", "System.Core", "Microsoft.CSharp");

        var cs = CSharpScript.Create(code, so);
        var compilation = cs.GetCompilation();
        using MemoryStream ms = new MemoryStream();
        var rslt = compilation.Emit(ms);
        // ==== finish preparing ====

        while (true)
        {
            ms.Seek(0, SeekOrigin.Begin);

            AssemblyLoadContext lc = new AssemblyLoadContext("test", isCollectible: true);
            var ass = lc.LoadFromStream(ms);

            var typ = ass.GetType("Submission#0");
            var mem = typ.GetMethod("<Factory>", BindingFlags.Static | BindingFlags.Public);

            var retTask = mem.Invoke(null, new object[] { new object[2] }) as Task<object>;
            var rsltTsk = await retTask;

            lc.Unload();

            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();

            await Task.Delay(200);
        }
    }
}
```

### Expected behavior

The output of the count will only increase once for the first time.

### Actual behavior

The output of the count always increase.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

**Which version of .NET is the code running on?**
.Net 5、.Net 6

**What OS and version, and what distro if applicable?**
Only Tested in Win 10 Pro.

**What is the architecture (x64, x86, ARM, ARM64)?**
x64

**Do you know whether it is specific to that configuration?**
No

### Other information

_No response_

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
`Microsoft.CSharp` keeps a cache of equivalent `ICSharpBinder`s and before returning one, it runs the binder through this cache, maybe replacing it with its cached equivalent. We can't prevent the allocation -it already happened-, but we manage reduce the number of long-lived objects with this trick.

The problem with this cache is that it keeps its members alive, and they contain `Type` objects, which means that if a method in an unloadable assembly that is using `dynamic` gets called, the assembly will become unable to be unloaded. This has been reported twice in #71629 and #78952.

This PR fixes the problem. The cache's global `ConcurrentDicrectory` was split into one `ConcurrentDirectory` per `AssemblyLoadContext`. The ALC is mapped to its cache with the help of a `ConditionalWeakTable`, which will let the GC collect the cache if the ALC gets unloaded. I added a test to validate that `dynamic` does not inhibit unloadability.

I think the change is low-risk; the main change is simple to follow and concentrated in the `BinderEquivalence` class and the rest of the library source code changes consist of just adding a parameter to a method call. I also think that given that the issue has been reported twice and there was no indication from the team that they don't want to fix it, and that it blocks unloading assemblies that use a particular language feature, it might be useful to merge despite the library's archived status.

Fixes #71629

### CODE REVIEW

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/BinderEquivalence.cs [hunk 1]
```diff
@@ -26,16 +28,27 @@ internal static class BinderEquivalence
...
+        private static readonly ConditionalWeakTable<AssemblyLoadContext,
+            ConcurrentDictionary<ICSharpBinder, ICSharpBinder>> binderEquivalenceCache =
+            new ConditionalWeakTable<AssemblyLoadContext, ConcurrentDictionary<ICSharpBinder, ICSharpBinder>>();
 
-        internal static T TryGetExisting<T>(this T binder)
+        internal static T TryGetExisting<T>(this T binder, Type context)
             where T : ICSharpBinder
         {
-            var fromCache = binderEquivalenceCache.GetOrAdd(binder, binder);
+            var alc = AssemblyLoadContext.GetLoadContext(context.Assembly);
```
Comment (jkotas) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/BinderEquivalence.cs:39]: This won't work well for non-collectible generic types instantiated over collectible types. E.g. consider how this is going to work for `List<MyCollectibleType>`. `context.Assembly` is going to be `System.Private.CoreLib`, so this type is going to be cached against the default non-unloadable load context that is not correct.

Comment (teo-tsirpanis) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/BinderEquivalence.cs:39]: Hmm, how about we just don't cache the binders if the context type `IsCollectible`?

**Comment by teo-tsirpanis** [comment]
After more investigations, there are multiple roots that inhibit unloadability. Here's a graph of them, this PR fixes the last one.

image

**Comment by cston** [comment]
@teo-tsirpanis, thanks for investigating this issue and providing a PR.

Unfortunately, we're not accepting significant changes to this library currently, to minimize the risk of regression, since the library is a mature code base that is no longer evolving. See [README.md](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.CSharp/README.md). Thanks.

**Pull request closed without merging**